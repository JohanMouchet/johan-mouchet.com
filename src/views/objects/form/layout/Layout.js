// @flow

import * as React from "react";
import cx from "classnames";
import "./Layout.scss";

type FieldsetProps = {
  legend?: string,
  className?: string | Array<string> | Object,
  children?: Node,
};

const Fieldset = ({ legend, className, children }: FieldsetProps) => {
  const classNames = cx("o-form__fieldset", className);

  return (
    <fieldset className={classNames}>
      <legend class="o-form__legend">{legend}</legend>
      {children}
    </fieldset>
  );
};

export { Fieldset };

type FormGroupProps = {
  variant?: "success" | "error" | "warning",
  detail?: string,
  children?: Node,
};

const FormGroup = ({ variant, detail, children }: FormGroupProps) => {
  const classNames = cx(
    "o-form__group",
    variant && `o-form__group--${variant}`
  );

  return (
    <div className={classNames}>
      {children}
      <span className="o-form__group-detail">{detail}</span>
    </div>
  );
};

export { FormGroup };

type LabelProps = {
  id?: string,
  block: boolean,
  className?: string | Array<string> | Object,
  children?: Node,
};

const Label = ({ id, block, className, children }: LabelProps) => {
  const classNames = cx(
    "o-form__label",
    (block || !children) && "o-form__label--block",
    className
  );

  return (
    <label htmlFor={id} className={classNames}>
      {children || "&nbsp;"}
    </label>
  );
};

export { Label };
