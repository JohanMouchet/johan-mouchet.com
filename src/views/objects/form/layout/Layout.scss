@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.o-form {
  $root: &;

  /* Fieldset */
  &__fieldset {
    border: 1px solid $s-color-gray;
    margin: t-vertical-rhythm(2) 0;
    padding: t-vertical-rhythm(1.5);
  }

  /* Legend */
  &__legend {
    margin-left: -0.5em;
    padding: 0 0.5em;
    color: $s-color-primary;
    text-transform: uppercase;
    font-weight: 500;
    font-size: t-modular-scale(0);
  }

  /* Form group */
  &__group {
    margin-bottom: t-vertical-rhythm(1);

    /** States **/

    /*** Success ***/
    &--success #{$root}__label {
      color: $s-color-success;
    }

    /*** Error ***/
    &--error #{$root}__label {
      color: $s-color-danger;
    }

    /*** Warning ***/
    &--warning #{$root}__label {
      color: $s-color-warning;
    }

    /** Alert **/
    &--success,
    &--error,
    &--warning {
      .o-alert {
        margin: t-vertical-rhythm(0.5) 0 0;
      }
    }
  }

  /* Label */
  &__label {
    display: inline-block;
    margin-top: t-vertical-rhythm(0.25);
    margin-bottom: t-vertical-rhythm(0.25);
    font-weight: 500;
    transition: color 0.3s;

    /** Block **/
    &--block {
      display: block;
    }
  }
}
