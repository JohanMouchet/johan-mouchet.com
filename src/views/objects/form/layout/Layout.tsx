import clsx, { ClassValue } from "clsx";
import styles from "./Layout.module.scss";

export const Fieldset = ({
  legend,
  children,
  className,
  ...props
}: {
  legend?: string;
  children?: React.ReactNode;
  className?: ClassValue;
} & React.HTMLProps<HTMLFieldSetElement>) => {
  return (
    <fieldset
      className={clsx(styles["o-form__fieldset"], className)}
      {...props}
    >
      {legend && <legend className={styles["o-form__legend"]}>{legend}</legend>}
      {children}
    </fieldset>
  );
};

export const FormGroup = ({
  variant,
  detail,
  children,
  className,
  ...props
}: {
  variant?: "success" | "error" | "warning";
  detail?: string;
  children?: React.ReactNode;
  className?: ClassValue;
} & React.HTMLProps<HTMLDivElement>) => {
  const classNames = clsx(
    styles["o-form__group"],
    variant && styles[`o-form__group--${variant}`],
    className
  );

  return (
    <div className={classNames} {...props}>
      {children}
      <span className={styles["o-form__group-detail"]}>{detail}</span>
    </div>
  );
};

export const Label = ({
  id,
  block,
  children,
  className,
  ...props
}: {
  id?: string;
  block?: boolean;
  children?: React.ReactNode;
  className?: ClassValue;
} & React.HTMLProps<HTMLLabelElement>) => {
  const classNames = clsx(
    styles["o-form__label"],
    (block || !children) && styles["o-form__label--block"],
    className
  );

  return (
    <label htmlFor={id} className={classNames} {...props}>
      {children || <>&nbsp;</>}
    </label>
  );
};
