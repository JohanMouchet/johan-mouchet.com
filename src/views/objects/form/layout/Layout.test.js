import React from "react";
import { shallow } from "enzyme";
import Layout from ".";

it("Renders <Layout /> without crashing", () => {
  shallow(<Layout />);
});
