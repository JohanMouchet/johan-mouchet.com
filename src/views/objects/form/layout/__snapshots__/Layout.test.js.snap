// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Fieldset /> without crashing 1`] = `
<fieldset
  className="o-form__fieldset"
>
  <legend
    class="o-form__legend"
  />
</fieldset>
`;

exports[`Renders <FormGroup /> without crashing 1`] = `
<div
  className="o-form__group"
>
  <span
    className="o-form__group-detail"
  />
</div>
`;

exports[`Renders <Label /> without crashing 1`] = `
<label
  className="o-form__label o-form__label--block"
>
  &nbsp;
</label>
`;
